### PR TITLE
general tracing context interface

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <comems@msn.com>

Sub-PR of #16049. Check #16049 get more background information.

This PR designed a new generic abstraction `TraceContext` to replace `Http::RequestHeaderMap` to provide tracing context to the tracer driver. `Http::RequestHeaderMapImpl` already inherits from TraceContext and implements the relevant interfaces.

Next, we just need simply replace `Http::RequestHeaderMap` with `TraceContext` in all tracer drivers implementations. After that, the main body of the entire general tracing system is completed. I'm not sure, whether the replacement step requires a separate PR. Because it does not involve any new logic, but there will be a lot of file changes. 

Risk Level: Low.
Testing: Added.
Docs Changes: N/A
Release Notes:  N/A

### CODE REVIEW

**Comment by wbpcode** [comment]
@lizan @mattklein123  This PR introduces the new TraceContext. All that remains is to replace the Http::ReqeustHeaderMap with the TraceContext. 😃

**Comment by wbpcode** [comment]
@jmarantz  Here I completed the new TraceContext without LowerCaseString. 😄

File: include/envoy/tracing/trace_driver.h [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 #include "envoy/common/pure.h"
 #include "envoy/stream_info/stream_info.h"
+#include "envoy/tracing/trace_context.h"
```
Comment (rojkov) [inline include/envoy/tracing/trace_driver.h:9]: Is this really needed?

**Review by rojkov — COMMENTED** [review]
Thank you! Looks very good. I've added a couple of question.

File: test/test_common/utility.h [hunk 2]
```diff
@@ -843,6 +843,33 @@ class ConditionalInitializer {
   bool ready_ ABSL_GUARDED_BY(mutex_){false};
 };
 
+namespace Tracing {
+
+class TestTraceContextImpl : public Tracing::TraceContext {
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply test/test_common/utility.h:848]: Yes, the purpose of this class is to facilitate other tests that require TraceContext. So although there are no tests using it yet, I still provide this class in this PR.

File: include/envoy/tracing/trace_driver.h (hunk 1 above)
Comment (wbpcode) [reply include/envoy/tracing/trace_driver.h:9]: Yes, for now this is still not needed and I will remove it.

File: source/common/http/header_map_impl.h [hunk 3]
```diff
@@ -480,6 +483,14 @@ class RequestHeaderMapImpl final : public TypedHeaderMapImpl<RequestHeaderMap>,
   INLINE_REQ_RESP_STRING_HEADERS(DEFINE_INLINE_HEADER_STRING_FUNCS)
   INLINE_REQ_RESP_NUMERIC_HEADERS(DEFINE_INLINE_HEADER_NUMERIC_FUNCS)
 
+  absl::optional<absl::string_view> getTraceContext(absl::string_view key) const override;
+
+  void setTraceContext(absl::string_view key, absl::string_view val) override;
+
+  void setTraceContextReferenceKey(absl::string_view key, absl::string_view val) override;
+
+  void setTraceContextReference(absl::string_view key, absl::string_view val) override;
```
Comment (rojkov) [inline source/common/http/header_map_impl.h:492]:
```suggestion
  // Tracing::TraceContext
  absl::optional<absl::string_view> getTraceContext(absl::string_view key) const override;
  void setTraceContext(absl::string_view key, absl::string_view val) override;
  void setTraceContextReferenceKey(absl::string_view key, absl::string_view val) override;
  void setTraceContextReference(absl::string_view key, absl::string_view val) override;
```
Sorry, I forgot to add this suggestion last time. Usually we add a comment with a reference to the base class where the overridden methods come from.

Comment (wbpcode) [reply source/common/http/header_map_impl.h:492]: Yes. This is necessary. It was an oversight on my part. 👍

**Review by rojkov — APPROVED** [review]
Thanks! LGTM. @lizan I think this is ready for a second pass.

File: source/common/http/header_map_impl.cc [hunk 4]
```diff
@@ -650,5 +658,49 @@ HeaderMapImplUtility::getAllHeaderMapImplInfo() {
...
+void RequestHeaderMapImpl::setTraceContext(absl::string_view key, absl::string_view val) {
+  ASSERT(validatedLowerCaseString(key));
+  HeaderMapImpl::removeExisting(key);
+
+  HeaderString new_key;
+  new_key.setCopy(key);
+  HeaderString new_val;
+  new_val.setCopy(val);
+
+  HeaderMapImpl::insertByKey(std::move(new_key), std::move(new_val));
```
Comment (lizan) [inline source/common/http/header_map_impl.cc:681]: no need to move string_view, it does nothing.

File: source/common/http/header_map_impl.cc [hunk 5]
```diff
@@ -40,6 +40,12 @@ InlineHeaderVector& getInVec(VariantHeader& buffer) {
 const InlineHeaderVector& getInVec(const VariantHeader& buffer) {
   return absl::get<InlineHeaderVector>(buffer);
 }
+
+bool validatedLowerCaseString(absl::string_view str) {
+  auto lower_case_str = LowerCaseString(std::string(str));
```
Comment (lizan) [inline source/common/http/header_map_impl.cc:45]: We might want to make LowerCaseString constructor takes string_view.

File: test/test_common/utility.h [hunk 6]
```diff
@@ -843,6 +843,33 @@ class ConditionalInitializer {
...
+public:
+  TestTraceContextImpl(const std::initializer_list<std::pair<std::string, std::string>>& values) {
+    for (const auto& value : values) {
+      context_map_[value.first] = value.second;
+    }
+  }
+
+  absl::optional<absl::string_view> getTraceContext(absl::string_view key) const override {
+    auto iter = context_map_.find(std::string(key));
```
Comment (lizan) [inline test/test_common/utility.h:857]:
```suggestion
    auto iter = context_map_.find(key);
```

absl::flat_hash_map support heterogeneous lookup

File: source/common/http/header_map_impl.cc (hunk 4 above)
Comment (wbpcode) [reply source/common/http/header_map_impl.cc:681]: 🤣 Type of `new_key/new_val` is `Http::HeaderString` rather then string_view and `HeaderMapImpl::insertByKey` use `Http::HeaderString&&` as parameter type. It seems that this move is necessary.

```
void HeaderMapImpl::insertByKey(HeaderString&& key, HeaderString&& value) {
  // ......
}

```

File: test/test_common/utility.h (hunk 6 above)
Comment (wbpcode) [reply test/test_common/utility.h:857]: get it.

File: source/common/http/header_map_impl.cc (hunk 5 above)
Comment (wbpcode) [reply source/common/http/header_map_impl.cc:45]: get it.

File: source/common/http/header_map_impl.cc [hunk 7]
```diff
@@ -650,5 +658,49 @@ HeaderMapImplUtility::getAllHeaderMapImplInfo() {
...
+  HeaderMapImpl::removeExisting(key);
+
+  HeaderString new_key;
+  new_key.setCopy(key);
+  HeaderString new_val;
+  new_val.setCopy(val);
+
+  HeaderMapImpl::insertByKey(std::move(new_key), std::move(new_val));
+  ASSERT(new_key.empty()); // NOLINT(bugprone-use-after-move)
+  ASSERT(new_val.empty()); // NOLINT(bugprone-use-after-move)
```
Comment (lizan) [inline source/common/http/header_map_impl.cc:683]: you don't need to ASSERT move behavior

**Review by lizan — COMMENTED** [review]
LGTM, just one nit

File: source/common/http/header_map_impl.cc (hunk 7 above)
Comment (wbpcode) [reply source/common/http/header_map_impl.cc]: get it.

**Comment by wbpcode** [comment]
@lizan ASSERT removed and main merged.

File: source/common/http/header_map_impl.cc [hunk 8]
```diff
@@ -650,5 +658,46 @@ HeaderMapImplUtility::getAllHeaderMapImplInfo() {
   return ret;
 }
 
+absl::optional<absl::string_view>
+RequestHeaderMapImpl::getTraceContext(absl::string_view key) const {
+  ASSERT(validatedLowerCaseString(key));
+  auto result = const_cast<RequestHeaderMapImpl*>(this)->getExisting(key);
```
Comment (lizan) [inline source/common/http/header_map_impl.cc:664]: The `const_cast` doesn't looks appropriate here. Why `getExisting` has to be non-const?

Comment (wbpcode) [reply source/common/http/header_map_impl.cc:664]: @lizan This is the original design of `HeaderMapImpl`. `HeaderMapImpl` also uses `const_cast` when implementing the 
original `get` method.
```
HeaderMap::GetResult HeaderMapImpl::get(const LowerCaseString& key) const {
  return HeaderMap::GetResult(const_cast<HeaderMapImpl*>(this)->getExisting(key));
}
```

> Why getExisting has to be non-const?

There are two main reasons I see.

First, the `staticLookup` that `getExisting` depends on is non-const. The reason `staticLookup` is non-const is that the `inlineHeaders` that `staticLookup` depends on is non-const.

Second, during the execution of the `getExsiting` method, `lazy_map_` may be initialized to speed up the search. This requires changing the state of HeaderMapImpl, so getExsiting needs to be non-const.

Comment (lizan) [reply source/common/http/header_map_impl.cc:664]: Fair enough, I think we want to make `lazy_map_` mutable, but that is out of scope of this PR.

**Review by lizan — APPROVED** [review]

**Pull request merged**